### PR TITLE
Missing meta

### PR DESCRIPTION
**Pull request by 928PJY**
#2813

### CODE REVIEW

File: src/docfx/build/metadata/Metadata.cs [hunk 1]
```diff
@@ -21,5 +28,101 @@ where item.Match(file.FilePath)
 
             return JsonUtility.Merge(config.GlobalMetadata, fileMetadata);
         }
+
+        public static JObject GenerateRawMetadata(Context context, Document file, TableOfContentsMap tocMap, string html)
```
Comment (yufeih) [inline src/docfx/build/metadata/Metadata.cs]: Put these metadatas to PageModel and convert them to legacy format in LegacyMarkdown.cs:

- WordCount
- Locale
- Toc

File: src/docfx/config/Config.cs [hunk 2]
```diff
@@ -15,6 +15,16 @@ internal class Config
         private static readonly string[] s_defaultContentInclude = new[] { "docs/**/*.{md,yml,json}" };
         private static readonly string[] s_defaultContentExclude = Array.Empty<string>();
 
+        /// <summary>
+        /// Docset name.
+        /// </summary>
+        public readonly string DocsetName = string.Empty;
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: We can have just a single config `Name` and use `{Product}.{DocsetName}` convension

File: src/docfx/build/metadata/Metadata.cs [hunk 3]
```diff
@@ -21,5 +28,101 @@ where item.Match(file.FilePath)
...
+            if (string.IsNullOrWhiteSpace(text))
+            {
+                return 0;
+            }
+
+            string[] wordList = text.Split(s_delimChars, StringSplitOptions.RemoveEmptyEntries);
+            return wordList.Count(s => !s.Trim().All(SpecialChars.Contains));
+        }
+
+        private static string GetRawTitle(string html)
```
Comment (yufeih) [inline src/docfx/build/metadata/Metadata.cs]: Title is already extracted in `BuildMarkdown`, can rawTitle be generated from `title`?

File: src/docfx/build/metadata/Metadata.cs [hunk 4]
```diff
@@ -21,5 +28,101 @@ where item.Match(file.FilePath)
...
+            // TODO: word count does not work for CJK locales...
+            if (html == null)
+            {
+                throw new ArgumentNullException(nameof(html));
+            }
+
+            HtmlDocument document = new HtmlDocument();
+
+            // Append a space before each end bracket so that InnerText inside different child nodes can separate itself from each other.
+            document.LoadHtml(html.Replace("</", " </", StringComparison.OrdinalIgnoreCase));
```
Comment (yufeih) [inline src/docfx/build/metadata/Metadata.cs]: we should only load html once.

File: src/docfx/build/metadata/Metadata.cs (hunk 3 above)
Comment (928PJY) [reply src/docfx/build/metadata/Metadata.cs]: `title` lose the `outerHtml` information

---
In reply to: [192974604](https://github.com/dotnet/docfx/pull/2840#discussion_r192974604) [](ancestors = 192974604)

File: src/docfx/build/metadata/Metadata.cs (hunk 4 above)
Comment (928PJY) [reply src/docfx/build/metadata/Metadata.cs]: in word counting, we have to do some replacement to the html, which can not be shared with others.

---
In reply to: [192974761](https://github.com/dotnet/docfx/pull/2840#discussion_r192974761) [](ancestors = 192974761)

File: src/docfx/build/metadata/Metadata.cs [hunk 5]
```diff
@@ -21,5 +28,101 @@ where item.Match(file.FilePath)
...
             return JsonUtility.Merge(config.GlobalMetadata, fileMetadata);
         }
+
+        public static JObject GenerateRawMetadata(Context context, Document file, TableOfContentsMap tocMap, string html)
+        {
+            var rawMetadata = new JObject();
+            var locale = file.Docset.Config.Locale;
+            var depotName = $"{file.Docset.Config.Product}{file.Docset.Config.DocsetName}";
+
+            rawMetadata["_op_canonicalUrlPrefix"] = $"https://docs.microsoft.com/{file.Docset.Config.Locale}/{file.Docset.Config.SiteBasePath}";
```
Comment (superyyrrzz) [inline src/docfx/build/metadata/Metadata.cs]: >docs.microsoft.com [](start = 62, length = 18)

make this configurable hosts isn't fixed?

File: src/docfx/build/metadata/Metadata.cs [hunk 6]
```diff
@@ -21,5 +28,101 @@ where item.Match(file.FilePath)
...
+            rawMetadata["_op_pdfUrlPrefixTemplate"] = $"https://docs.microsoft.com/pdfstore/{locale}/{depotName}/{{branchName}}{{pdfName}}";
+            rawMetadata["_op_rawTitle"] = GetRawTitle(html);
+
+            rawMetadata["_op_wordCount"] = rawMetadata["word_count"] = CountWord(html);
+
+            rawMetadata["depot_name"] = depotName;
+            rawMetadata["is_dynamic_rendering"] = true;
+            rawMetadata["layout"] = "Conceptual";
```
Comment (superyyrrzz) [inline src/docfx/build/metadata/Metadata.cs]: >rawMetadata["layout"] = "Conceptual"; [](start = 12, length = 37)

use the value user set in metadata if set

**Comment by superyyrrzz** [comment]
what about add a test spec for legacy?

File: src/docfx/docfx.csproj [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
```
Comment (yufeih) [inline src/docfx/docfx.csproj]: >0 [](start = 33, length = 1)

2.1

File: src/docfx/build/legacy/LegacyMarkdown.cs [hunk 8]
```diff
@@ -21,12 +21,19 @@ internal static class LegacyMarkdown
...
-                pageModel.Content = PostProcessHtml(pageModel.Content, docset.Config.Locale);
+                legacyRawMetadata.Content = PostProcessHtml(pageModel.Content, docset.Config.Locale);
             }
 
-            context.WriteJson(pageModel, rawPageOutputPath);
+            legacyRawMetadata.RawMetadata = pageModel.Metadata;
+            legacyRawMetadata.RawMetadata.Metadata["toc_rel"] = pageModel.TocRelativePath;
+            legacyRawMetadata.RawMetadata.Metadata["locale"] = pageModel.Locale;
+            legacyRawMetadata.RawMetadata.Metadata["word_count"] = pageModel.WordCount;
+            legacyRawMetadata.RawMetadata.Metadata["_op_rawTitle"] = $"<h1>{pageModel.Metadata.Title}</h1>";
```
Comment (yufeih) [inline src/docfx/build/legacy/LegacyMarkdown.cs]: >pageModel.Metadata.Title [](start = 76, length = 24)

html escape?

File: src/docfx/lib/HtmlUtility.cs [hunk 9]
```diff
@@ -24,6 +29,26 @@ public static void AddLinkType(HtmlNode html, string locale)
             AddLinkType(html, "img", "src", locale);
         }
 
+        public static long CountWord(HtmlNode html)
```
Comment (yufeih) [inline src/docfx/lib/HtmlUtility.cs]: >CountWord [](start = 27, length = 9)

add tests

File: src/docfx/build/metadata/Metadata.cs [hunk 10]
```diff
@@ -21,5 +23,24 @@ where item.Match(file.FilePath)
 
             return JsonUtility.Merge(config.GlobalMetadata, fileMetadata);
         }
+
+        public static JObject GenerateRawMetadata(Document file, HtmlNode html, string locale, long wordCount)
```
Comment (yufeih) [inline src/docfx/build/metadata/Metadata.cs]: >GenerateRawMetadata [](start = 30, length = 19)

Move it to legacy

File: src/docfx/config/Config.cs [hunk 11]
```diff
@@ -15,6 +15,11 @@ internal class Config
         private static readonly string[] s_defaultContentInclude = new[] { "docs/**/*.{md,yml,json}" };
         private static readonly string[] s_defaultContentExclude = Array.Empty<string>();
 
+        /// <summary>
+        /// Depot name. = _(ops `DepotName` = `<ProductName>.<DocsetName>`)_
+        /// </summary>
+        public readonly string Name = string.Empty;
```
Comment (herohua) [inline src/docfx/config/Config.cs]: >string.Empty [](start = 38, length = 12)

why is this string.Empty?

Comment (928PJY) [reply src/docfx/config/Config.cs]: it is a default value and will be assigned with the user config.

---
In reply to: [194420000](https://github.com/dotnet/docfx/pull/2840#discussion_r194420000) [](ancestors = 194420000)

File: src/docfx/lib/HtmlUtility.cs [hunk 12]
```diff
@@ -25,6 +35,31 @@ public static HtmlNode AddLinkType(this HtmlNode html, string locale)
             return html;
         }
 
+        public static long CountWord(HtmlNode html)
+        {
+            // TODO: word count does not work for CJK locales...
+
+            // To improve the performance, we just load html once, but the innerText will connect two element's innerText directly and will miss one.
+            // For example: <h1>test header</h2><p>test paragraph</p> => InnerText: test headertest paragraph
+            long wordCount = CountWordInText(html.InnerText);
```
Comment (superyyrrzz) [inline src/docfx/lib/HtmlUtility.cs]: >html.InnerText [](start = 45, length = 14)

add a TODO to change logic later to keep result stable

**Pull request merged**